### ISSUE TITLE
Feature Request (Typescript): Improve System and Component creation methods to avoid misconfiguration in Typescript

### ISSUE THREAD

**Issue by cfortuner** [MozillaReality/ecsy#187]
HI!

Loving the library thanks for all the hard work! 

I'm wondering if it might be better to change the way systems and components are registered and added to make it easy to enforce types when configuring them.

**Problem**: 
Today, when I want to register a system with attributes, I pass the system to the world.registerSystem function with an attributes map. This is prone to error as there is no way (I can think of without wrapping this call in helper methods) to enforce that the attributes map contains the correct attributes that the system needs.

Many systems I'm creating do need attributes to work and yet it feels very easy to misconfigure them.

I have the same issue with Components. Components that need attributes provided to them don't have a way of enforcing that they are provided the correct types. Another frustrating issue with components in typescript, is that any component property must be marked with a default value or undefined. While I understand that components should be data, it can be much easier to work with if I relax that constraint and allow components to have more complex attributes (like a game object from the engine I'm using). The problem is, in order to create non trivial default values in the component, external systems must be accessed, which encourages me to use globals. I've found a way to provide them through attributes, but like with the world.registerSystem call, it is prone to errors as there is no clear way to enforce what the attributes map should be.

I could be way off here or missing something so please don't hesitate to let me know if there is a better way to do this or if I'm not thinking about this tool in the right way!

**Proposal**:
I would like to propose that Systems and Components both get new static constructor methods that help enforce creation with the correct attributes.

Something along the lines of:
```
System: 
static register(world: World, attributes: Attributes) => System {
   world.registerSystem(System, attributes)
}

Component:
static register(entity: Entity, attributes: Attributes) => Component {
   entity.addComponent(entity, attributes)
}
```

These methods can just wrap the world.registerSystem(System, attributes) and entity.addComponent(component: Component, attributes: Attributes) methods but they allow for the System and Component attributes to be enforced like so:

```
interface MySystemAttributes extends Attributes {}

class MySystem extends System {
static init(world: World, attributes: MySystemAttributes
```

Of course I can just extend System and Component myself and add these  but I thought I would post to see what the community thinks!  

Thanks!
Colin

**Comment by cfortuner** [comment]
Ah I see this pr might address some of my concerns https://github.com/MozillaReality/ecsy/pull/168/files and enforce that components are configured correcting in JS.

**Comment by fernandojsg** [comment]
@cfortuner we just merged that PR #168 and the #191 
does it solve your problem?

**Comment by fernandojsg** [comment]
Closing it now as I believe all the typing issues has been solved already with the recent PRs by @robertlong
